File: tutorials/distributed_data_cache/c++11/interactive_sensor_publisher.cxx
//
// (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
//
//  RTI grants Licensee a license to use, modify, compile, and create derivative
//  works of the Software.  Licensee has the right to distribute object form
//  only for use with RTI products.  The Software is provided "as is", with no
//  warranty of any type, including any warranty for fitness for any purpose.
//  RTI is under no obligation to maintain or support the Software.  RTI shall
//  not be liable for any incidental or consequential damages arising out of the
//  use or inability to use the software.
//

#include <thread>
#include "rti/rti.hpp"
#include "home_automation.hpp"

class WindowSensor {
  public:
    WindowSensor(const std::string& sensor_name, const std::string& room_name)
            : status_(sensor_name, room_name, false),
            instance_(dds::core::InstanceHandle::nil()),
            writer_(nullptr)
    {
        dds::domain::DomainParticipant participant(0);
        dds::topic::Topic<DeviceStatus> topic(participant, "WindowStatus");
        writer_ = std::make_shared<dds::pub::DataWriter<DeviceStatus>>(
                dds::pub::DataWriter<DeviceStatus>(topic));

        instance_ = writer_->register_instance(status_);
        writer_->write(status_, instance_);
    }

    void open_window()
    {
        std::cout << "Opening the window..." << std::endl;
        status_.is_open(true);
        writer_->write(status_, instance_);
    }

    void close_window()
    {
        std::cout << "Closing the window..." << std::endl;
        status_.is_open(false);
        writer_->write(status_, instance_);
    }

    void turn_off()
    {
        std::cout << "Turning off the sensor..." << std::endl;
        writer_->dispose_instance(instance_);
    }

    void turn_on()
    {
        std::cout << "Turning on the sensor..." << std::endl;
        writer_->write(status_, instance_);
    }

  private:
    DeviceStatus status_;
    dds::core::InstanceHandle instance_;
    std::shared_ptr<dds::pub::DataWriter<DeviceStatus>> writer_;

};

void control_sensor(
        const std::string& sensor_name,
        const std::string& room_name)
{
    WindowSensor window_sendor(sensor_name, room_name);
    std::string option;

    while (true) {
        std::cout << "\nEnter one of the following options:" << std::endl
            << "\t- open" << std::endl
            << "\t- close" << std::endl
            << "\t- off" << std::endl
            << "\t- on" << std::endl
            << "\t- exit" << std::endl;

        std::cin >> option;

        if (option == "open") {
            window_sendor.open_window();
        } else if (option == "close") {
            window_sendor.close_window();
        } else if (option == "off") {
            window_sendor.turn_off();
        } else if (option == "on") {
            window_sendor.turn_on();
        } else if (option == "exit") {
            break;
        } else {
            std::cout << "Invalid option, please enter a valid option [open|close|off|on|exit]." << std::endl;
        }
    }
}

int main(int argc, char **argv)
{
    std::string sensor_name = (argc > 1) ? argv[1] : "Sensor1";
    std::string room_name = (argc > 2) ? argv[2] : "LivingRoom";
    control_sensor(sensor_name, room_name);
}

File: tutorials/distributed_data_cache/py/interactive_sensor_publisher.py
#
# (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
#
# RTI grants Licensee a license to use, modify, compile, and create derivative
# works of the Software solely for use with RTI products.  The Software is
# provided "as is", with no warranty of any type, including any warranty for
# fitness for any purpose. RTI is under no obligation to maintain or support
# the Software.  RTI shall not be liable for any incidental or consequential
# damages arising out of the use or inability to use the software.
#

import sys
from time import sleep

import rti.connextdds as dds
from home_automation import DeviceStatus


class WindowSensor:

    def __init__(self, sensor_name: str, room_name: str) -> None:
        self.status = DeviceStatus(sensor_name, room_name, is_open=False)

        participant = dds.DomainParticipant(domain_id=0)
        topic = dds.Topic(participant, "WindowStatus", DeviceStatus)
        self.writer = dds.DataWriter(topic)

        self.instance = self.writer.register_instance(self.status)
        self.writer.write(self.status, self.instance)

    def open_window(self) -> None:
        print("Opening the window...")
        self.status.is_open = True
        self.writer.write(self.status, self.instance)

    def close_window(self) -> None:
        print("Closing the window...")
        self.status.is_open = False
        self.writer.write(self.status, self.instance)

    def turn_off(self) -> None:
        print("Turning off the sensor...")
        self.writer.dispose_instance(self.instance)

    def turn_on(self) -> None:
        print("Turning on the sensor...")
        self.writer.write(self.status, self.instance)


def control_sensor(sensor_name: str, room_name: str):

    window_sensor = WindowSensor(sensor_name, room_name)
    help_message = """\nEnter one of the following options:
        - open
        - close
        - off
        - on
        - exit\n"""

    while True:
        value = input(help_message).lower()

        if value == "open":
            window_sensor.open_window()
        elif value == "close":
            window_sensor.close_window()
        elif value == "off":
            window_sensor.turn_off()
        elif value == "on":
            window_sensor.turn_on()
        elif value == "exit":
            break
        else:
            print(
                "Invalid option, please enter a valid option [open|close|off|on|exit]."
            )


if __name__ == "__main__":
    sensor_name = sys.argv[1] if len(sys.argv) > 1 else "Window1"
    room_name = sys.argv[2] if len(sys.argv) > 2 else "LivingRoom"

    try:
        control_sensor(sensor_name, room_name)
    except KeyboardInterrupt:
        pass
